services/auth/reverseproxy.go

```diff
@@ -55,32 +57,74 @@ func (r *ReverseProxy) Name() string {
 // the revese proxy.
 // If a username is available in the "setting.ReverseProxyAuthUser" header an existing
 // user object is returned (populated with username or email found in header).
-// Returns nil if header is empty.
+// Returns nil if header is empty or internal API is being called.
 func (r *ReverseProxy) Verify(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *user_model.User {
+
+	// Internal API should not use this auth method.
+	if middleware.IsInternalPath(req) {
+		return nil
+	}
+
+	// Just return user if session is estabilshed already.
+	user := SessionUser(sess)
+	if user != nil {
+		return user
+	}
```

Comment:
This mixes the `Session` provider into the reverse proxy. I think we should not do this. Is it needed for the new logic?

Reply:
> This mixes the `Session` provider into the reverse proxy. I think we should not do this. Is it needed for the new logic?

Reverse proxy is/should be for user authentication stuff not Gitea's session handling IHMO. So there is no point in user verification if already verified (established session). Don't see problem with it.

Reply:
Then it's just a wrong execution order and the session provider should be checked first.

Reply:
`SessionUser(sess)` is called first and used if session is already estabilshed; if no session is established, new session is created. Don't see any problem here. If you see problem here, please propose exact code fix.

Reply:
The session handler is already called before the reverse proxy handler:
https://github.com/go-gitea/gitea/blob/cb10f2767c920fdbca6fc43e71dcc4c4b8c6e0d0/routers/web/web.go#L104-L111

Reply:
Fixed in dfc544c401c2674886ec40c6587997d5a9337b56. Thanks for pointing.